src/optimagic/parameters/scale_conversion.py

```diff
@@ -1,30 +1,65 @@
-from functools import partial
-from typing import Callable, NamedTuple
+from dataclasses import dataclass
 
 import numpy as np
+from numpy.typing import NDArray
 
+from optimagic.parameters.scaling import ScalingOptions
 from optimagic.parameters.space_conversion import InternalParams
 
 
+@dataclass(frozen=True)
+class ScaleConverter:
+    factor: NDArray[np.float64] | None
+    offset: NDArray[np.float64] | None
+
+    def params_to_internal(self, vec: NDArray[np.float64]) -> NDArray[np.float64]:
+        """Scale a parameter vector from external scale to internal one."""
+        if self.offset is not None:
+            vec = vec - self.offset
+        if self.factor is not None:
+            vec = vec / self.factor
+        return vec
+
+    def params_from_internal(self, vec: NDArray[np.float64]) -> NDArray[np.float64]:
+        """Scale a parameter vector from internal scale to external one."""
+        if self.factor is not None:
+            vec = vec * self.factor
+        if self.offset is not None:
+            vec = vec + self.offset
+        return vec
+
+    def derivative_to_internal(
+        self, derivative: NDArray[np.float64]
+    ) -> NDArray[np.float64]:
+        """Scale a derivative vector from external scale to internal one."""
+        if self.factor is not None:
+            derivative = derivative * self.factor
+        return derivative
+
+    def derivative_from_internal(
+        self, derivative: NDArray[np.float64]
+    ) -> NDArray[np.float64]:
+        """Scale a derivative vector from internal scale to external one."""
+        if self.factor is not None:
+            derivative = derivative / self.factor
+        return derivative
+
+
 def get_scale_converter(
-    internal_params,
-    scaling,
-    scaling_options,
-):
+    internal_params: InternalParams,
+    scaling: ScalingOptions | None,
+) -> tuple[ScaleConverter, InternalParams]:
     """Get a converter between scaled and unscaled parameters.
 
     Args:
-        internal_params (InternalParams): NamedTuple of internal and possibly
-            reparametrized but not yet scaled parameter values and bounds.
-        func (callable): The criterion function. Possibly used to calculate a scaling
-            factor.
-        scaling (bool): Whether scaling should be done.
-        scaling_options (dict): User provided scaling options.
+        internal_params: NamedTuple of internal and possibly reparametrized but not yet
+            scaled parameter values and bounds.
+        scaling: Scaling options. If None, no scaling is performed.
 
     Returns:
-        ScaleConverter: NamedTuple with methods to convert between scaled and unscaled
+        ScaleConverter: Dataclass with methods to convert between scaled and unscaled
             internal parameters and derivatives.
-        InternalParams: NamedTuple with entries:
+        InternalParams: Dataclass with entries:
```

Comment:
We don't need to list all the attributes of `InternalParams` as it should be documented by that class.